ext/node/path/posix.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::AnyError;
+use deno_core::op;
+use deno_core::serde_json::Value;
+
+use super::assert_path;
+use super::char_at_index;
+use super::constants;
+
+#[op]
+pub fn op_node_path_posix_dirname(path: Value) -> Result<String, AnyError> {
+  let path = assert_path(&path)?;
+  let char_count = path.chars().count();
+  if char_count == 0 {
+    return Ok(String::from("."));
+  }
+
+  let has_root =
+    char_at_index(path, 0).unwrap() as u32 == constants::CHAR_FORWARD_SLASH;
+  let mut end: Option<usize> = None;
+  let mut matched_slash = true;
+
+  for (i, c) in path.char_indices().rev() {
+    if i < 1 {
+      break;
+    }
+    match c as u32 {
+      constants::CHAR_FORWARD_SLASH if !matched_slash => {
+        end = Some(i);
+        break;
+      }
+      constants::CHAR_FORWARD_SLASH => {
+        // noop
+      }
+      _ => matched_slash = false,
+    }
+  }
+
+  if end.is_none() {
```

Comment:
would match on tuple be a bit cleaner?
```Rust
  match (end, has_root) {
    (None, true) => String::from("/"),
    (None, false) => String::from("."),
    (Some(1), true) => String::from("//"),
    (Some(ending), _) => path.chars().take(ending).collect(),
  }
```
@20k-ultra  feel free to ignore me I am far from expert at Rust.

Reply:
yeah I love that. I was so focused on doing a direct port I tried avoiding any optimizations but this is the same logically just more idiomatic.

Reply:
I was able to do this for win32.rs also

```
  let dirname = match (end, root_end) {
    (None, None) => String::from("."),
    (None, Some(e)) => path.chars().take(e).collect(),
    (Some(e), _) => path.chars().take(e).collect(),
  };

  Ok(dirname)
```